GH-43057: [C++] Thread-safe AesEncryptor / AesDecryptor

EnricoMi opened the pull request:
### Rationale for this change

OpenSSL encryption / decryption is wrapped by AesEncryptor / AesDencryptor, which is used by multiple threads of a single scanner or by multiple concurrent scanners when scanning a dataset. Some thread may call `WipeOut` while other threads still use the instance.

### What changes are included in this PR?

- Remove the `WipeOut` methods and related datastructures entirely.
- Each call into `CtrEncrypt` / `CtrDecrypt` and `GcmEncrypt` / `GcmDecrypt` uses its own `EVP_CIPHER_CTX` instance, making this thread-safe.
- To reduce initialization time of the `EVP_CIPHER_CTX` instance, a copy of an initalized but unused context is used.

After fixing this `"AesDecryptor was wiped out"` issue, two other segmentation faults surfaced: GH-44988. This has also been addressed here as it can only be exposed after fixing the wipe-out issue.

Fixes GH-43057.
Fixes GH-44852.
Fixes GH-44988.

### Are these changes tested?
A unit test that scans a dataset concurrently reproduced the initial issue in 30% of the test runs.

### Are there any user-facing changes?
No.

adamreeve commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,85 @@ constexpr int32_t kBufferSizeLength = 4;
+
+  /// Create a new cipher context that auto-frees
+  /// This duplicates un unused but initialized private context to avoid going through
+  /// initialization
```
Have you measured how long initialization takes and how long it takes to copy an initialized context? How much performance benefit does this approach provide over initializing a context each time one is needed?

pitrou replied:
I agree that ideally we wouldn't do this at all. I'd rather remove any optimization around this first, and re-optimize in later PRs if we find that cipher creation takes a significant amount of time (it sounds unlikely it will).

EnricoMi replied:
Took some measurements:
- creating, initializing and freeing a contexts takes 1.7µs
- duplicating and freeing a context takes 1.2µs

This was measured over 10,000,000 instances (17s and 12s).

I'll remove complexity given creating a context takes µs.

adamreeve commented on cpp/src/parquet/encryption/internal_file_decryptor.cc:
```diff
@@ -64,16 +64,6 @@ InternalFileDecryptor::InternalFileDecryptor(FileDecryptionProperties* propertie
-void InternalFileDecryptor::WipeOutDecryptionKeys() {
-  std::lock_guard<std::mutex> lock(mutex_);
-  properties_->WipeOutDecryptionKeys();
-  for (auto const& i : all_decryptors_) {
```
I think being able to wipe out the decryptors is the only reason the `all_decryptors_` and `mutex_` members exist, so those should be able to be removed.

EnricoMi replied:
thanks, done

adamreeve commented:
The test failures look like they might be caused by https://github.com/openssl/openssl/issues/21955

The failing ASAN test run is using OpenSSL 3.02, but builds with other versions are passing. Eg. the Ubuntu 20.04 build uses OpenSSL 1.1.1f and the Conda build has OpenSSL 3.3.2.

The fix for this was backported to the 3.0 branch and released in 3.0.13 (https://github.com/openssl/openssl/pull/23102), but Ubuntu 22.04 still uses 3.0.2. If there's not an easy way around this, maybe we need a slower code path for 3.0.2 that creates a new context each time?

pitrou commented:
> If there's not an easy way around this, maybe we need a slower code path for 3.0.2 that creates a new context each time?

Let's use the "slower" code path everywhere? I'm skeptical it will be that slow...

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,85 @@ constexpr int32_t kBufferSizeLength = 4;
+  virtual ~AesEncryptionContext() = default;
+
+ protected:
+  void InitCipherContext() {
```
Why not do this in the constructor?

EnricoMi replied:
The constructor cannot call into a virtual method implemented by derived classes. The constructor will call into the virtual method instead. Only after the constructor the overloading is available.

Moving away from duplicating the initialized context makes this method redundant any way.

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,85 @@ constexpr int32_t kBufferSizeLength = 4;
+    InitCipherContext(ctx_.get());
+  }
+
+  virtual void InitCipherContext(EVP_CIPHER_CTX* ctx) = 0;
```
This sort of indirection to the derived class looks convoluted, why not reverse the control flow and have the derived class call the base method before doing its own initialization?

EnricoMi replied:
removed

pitrou commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,85 @@ constexpr int32_t kBufferSizeLength = 4;
+
+  virtual void InitCipherContext(EVP_CIPHER_CTX* ctx) = 0;
+
+  std::function<void(EVP_CIPHER_CTX*)> ctxDeleter = [](EVP_CIPHER_CTX* ctx) {
```
Please let's stick to the coding conventions, e.g.:
```suggestion
  static inline std::function<void(EVP_CIPHER_CTX*)> ctx_deleter_ = [](EVP_CIPHER_CTX* ctx) {
```

(also can make it `static` since it doesn't need to be per instance)

EnricoMi replied:
done

pitrou commented on cpp/src/parquet/encryption/internal_file_decryptor.cc:
```diff
@@ -64,16 +64,6 @@ InternalFileDecryptor::InternalFileDecryptor(FileDecryptionProperties* propertie
 
-void InternalFileDecryptor::WipeOutDecryptionKeys() {
-  std::lock_guard<std::mutex> lock(mutex_);
-  properties_->WipeOutDecryptionKeys();
```
`properties_->WipeOutDecryptionKeys` will clear the private key, we should keep it no?
(even though it's currently not reliable: see https://github.com/apache/arrow/issues/31603)

EnricoMi replied:
restored

pitrou commented on cpp/src/parquet/encryption/internal_file_encryptor.cc:
```diff
@@ -50,21 +50,6 @@ InternalFileEncryptor::InternalFileEncryptor(FileEncryptionProperties* propertie
 }
 
-void InternalFileEncryptor::WipeOutEncryptionKeys() {
-  properties_->WipeOutEncryptionKeys();
```
Same here

EnricoMi replied:
dito

pitrou reviewed (changes requested):
Thanks a lot for stepping in and submitting this @EnricoMi . You'll find a number of comments below.

EnricoMi commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -544,7 +545,7 @@ Future<std::shared_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
                       std::move(const_cast<std::unique_ptr<parquet::ParquetFileReader>&>(
                           reader)),
-                      std::move(arrow_properties), &arrow_reader));
+                      arrow_properties, &arrow_reader));
```
Argument `const ArrowReaderProperties& properties` is passed as a const reference, so no move will happen:

    Clang-Tidy: Passing result of std::move() as a const reference argument; no move will actually happen

EnricoMi commented on cpp/src/parquet/encryption/encryption_internal.h:
```diff
@@ -44,6 +44,8 @@ constexpr int8_t kOffsetIndex = 7;
 constexpr int8_t kBloomFilterHeader = 8;
 constexpr int8_t kBloomFilterBitset = 9;
 
+class AesEncryptionContext;
```
removed

EnricoMi commented on cpp/src/parquet/encryption/encryption_internal.h:
```diff
@@ -44,6 +44,8 @@ constexpr int8_t kOffsetIndex = 7;
 constexpr int8_t kBloomFilterHeader = 8;
 constexpr int8_t kBloomFilterBitset = 9;
 
+class AesEncryptionContext;
```
removed

EnricoMi commented:
I have addressed all comments except for the test related ones. Will go over them tomorrow.

Thanks for your input, that is highly appreciated!

adamreeve commented:
I tested concurrent scans of a larger dataset with uniform encryption, and this change doesn't completely fix that scenario:

```diff
--- a/cpp/src/arrow/dataset/file_parquet_encryption_test.cc
+++ b/cpp/src/arrow/dataset/file_parquet_encryption_test.cc
@@ -289,6 +289,8 @@ TEST_F(DatasetEncryptionTest, ReadSingleFile) {
 // processing encrypted datasets over 2^15 rows in multi-threaded mode.
 class LargeRowEncryptionTest : public DatasetEncryptionTestBase {
  public:
+  LargeRowEncryptionTest() : DatasetEncryptionTestBase(true) {}
+
   // The dataset is partitioned using a Hive partitioning scheme.
   void PrepareTableAndPartitioning() override {
     // Specifically chosen to be greater than batch size for triggering prefetch.
@@ -307,7 +309,7 @@ class LargeRowEncryptionTest : public DatasetEncryptionTestBase {
 
 // Test for writing and reading encrypted dataset with large row count.
 TEST_F(LargeRowEncryptionTest, ReadEncryptLargeRows) {
-  ASSERT_NO_FATAL_FAILURE(TestScanDataset());
+  ASSERT_NO_FATAL_FAILURE(TestScanDataset(true));
 }
 
 }  // namespace dataset
```

This gives errors like:
```
failed with IOError: Failed decryption finalization
```

I believe this is due to the decryptor AAD being mutable. It's [updated for each data page read](https://github.com/apache/arrow/blob/6252e9ceeb0f8544c14f79d895a37ac198131f88/cpp/src/parquet/column_reader.cc#L357), so concurrent use will result in incorrect AADs being used in some threads.

Rather than mutating the decryptor state, it might be possible to refactor this so that the AAD values are passed in to the decrypt method as a parameter.

I think this should probably be addressed as a separate PR though as the changes should be orthogonal.

EnricoMi commented:
> failed with IOError: Failed decryption finalization

Confirmed in c3644aa43ad43a64a93b8c44bb88eef141fdae55. Multi-threaded uniform encryption read fails.

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -54,6 +55,9 @@ namespace dataset {
 class DatasetEncryptionTestBase : public ::testing::Test {
  public:
+  explicit DatasetEncryptionTestBase(bool uniform_encryption = false)
+      : uniform_encryption(uniform_encryption) {}
```
done

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -151,31 +159,56 @@ class DatasetEncryptionTestBase : public ::testing::Test {
   std::shared_ptr<parquet::encryption::CryptoFactory> crypto_factory_;
   std::shared_ptr<parquet::encryption::KmsConnectionConfig> kms_connection_config_;
+
+ private:
```
removed

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -151,31 +159,56 @@ class DatasetEncryptionTestBase : public ::testing::Test {
+    const size_t attempts = concurrently ? 1000 : 2;
+    for (size_t i = 0; i < attempts; ++i) {
+      if (concurrently) {
+        threads.push_back(std::async(DatasetEncryptionTestBase::read, dataset));
```
moved to Arrow Future

EnricoMi commented:
> > failed with IOError: Failed decryption finalization
> 
> Confirmed in [ec9b054](https://github.com/apache/arrow/commit/ec9b054dcf2b10bbed8c2c38b93b88a17d5d3d06). Multi-threaded uniform encryption read fails.

Attempt to fix multi-threaded AAD update: d77a081a49b3436fd0bc16ce3d10792b7222e026

The `PageReader` uses its own copy of the `data_decryptor` / `data_decryptor`. Happy to move into separate PR.

adamreeve commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -307,8 +307,15 @@ class SerializedPageReader : public PageReader {
   CryptoContext crypto_ctx_;
+  // This PageReader has its own copy of crypto_ctx_->meta_decryptor and
+  // crypto_ctx_->data_decryptor in order to be thread-safe. Do not mutate (update) the
+  // instances of crypto_ctx_.
```
There are also uses of `UpdateDecryptor` in `RowGroupPageIndexReaderImpl`. I'm not sure whether that's used concurrently anywhere within Arrow code, but it's part of the public API so could be used concurrently by users and it would make sense to make a similar change there.

wgtmac replied:
Yes, they are used only once for a specific column.

adamreeve replied:
But the decryptor returned by `GetColumnMetaDecryptor` might be the `InternalFileDecryptor::footer_metadata_decryptor_` member in the case of uniform encryption, so although it's only used once, trying to use this method concurrently for different columns would risk race conditions where the AAD might be updated in another thread between being updated and used.

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -51,8 +53,14 @@ using arrow::internal::checked_pointer_cast;
 namespace dataset {
 
+// Tests come in these variations
+enum CompressionParam {
```
you're right, fixed

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -90,7 +98,15 @@ class DatasetEncryptionTestBase : public ::testing::Test {
+    } else if (GetParam() == UNIFORM) {
+      encryption_config->uniform_encryption = true;
+    } else {
+      FAIL() << "Unsupported compression type " << GetParam();
```
fixed

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -151,21 +167,53 @@ class DatasetEncryptionTestBase : public ::testing::Test {
+      // ramp up parallelism
+      ASSERT_OK(pool->SetCapacity(16));
+      // ensure there are sufficient jobs to see concurrent processing
+      ASSERT_GT(pool->GetNumTasks(), 16);
```
these tests are flaky any way if they fail, this does not make it more deterministic, removed

EnricoMi commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -151,21 +167,53 @@ class DatasetEncryptionTestBase : public ::testing::Test {
+      ASSERT_OK(pool->SetCapacity(16));
+      // ensure there are sufficient jobs to see concurrent processing
+      ASSERT_GT(pool->GetNumTasks(), 16);
+      printf("%d", pool->GetNumTasks());
```
removed

pitrou commented on cpp/src/parquet/encryption/internal_file_decryptor.cc:
```diff
@@ -147,13 +141,10 @@ std::shared_ptr<Decryptor> InternalFileDecryptor::GetFooterDecryptor(
-    aes_data_decryptor = encryption::AesDecryptor::Make(
-        algorithm_, key_len, /*metadata=*/false, &all_decryptors_);
-  }
+  aes_metadata_decryptor =
+      encryption::AesDecryptor::Make(algorithm_, key_len, /*metadata=*/true);
+  aes_data_decryptor =
+      encryption::AesDecryptor::Make(algorithm_, key_len, /*metadata=*/false);
```
I still don't think this function is correct, because it still caches the decryptors. It's the footer key that should be cached in `InternalFileDecryptor`, not the decryptors themselves.

(also, please let's reuse `GetFooterKey` instead of duplicating its code)

EnricoMi replied:
Done! Removed code duplication of `GetFooterKey` and replaced caching of `aes_metadata_decryptor` and `aes_data_decryptor` with caching `footer_key`. Cached `footer_key` gets wiped out on `WipeOutDecryptionKeys`.

mapleFU commented:
Any updates?

pitrou commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -333,14 +340,16 @@ void SerializedPageReader::InitDecryption() {
   // Prepare the AAD for quick update later.
   if (crypto_ctx_.data_decryptor != nullptr) {
     ARROW_DCHECK(!crypto_ctx_.data_decryptor->file_aad().empty());
+    data_decryptor_ = std::make_shared<Decryptor>(*crypto_ctx_.data_decryptor);
```
So this is making a copy of the `Decryptor` object. But `Decryptor::aes_decryptor_` is a `std::shared_ptr<encryption::AesDecryptor>`, so this is merely going to increment the shared_ptr's reference count, right?

pitrou replied:
I think this means that we want to change:
```c++
std::shared_ptr<Decryptor> GetColumnDataDecryptor(
    const ColumnCryptoMetaData* crypto_metadata, InternalFileDecryptor* file_decryptor);
```
into:
```c++
std::function<std::shared_ptr<Decryptor>()> GetColumnDataDecryptor(
    const ColumnCryptoMetaData* crypto_metadata, InternalFileDecryptor* file_decryptor);
```
so that we can get a new Decryptor instance every time we need one.

(same for the meta decryptor, of course)

EnricoMi replied:
You are right,

    data_decryptor_ = std::make_shared<Decryptor>(*crypto_ctx_.data_decryptor)

could be simplified to

    data_decryptor_ = crypto_ctx_.data_decryptor

which should increment the reference counter and therefore be equivalent.

EnricoMi replied:
With `std::function<std::shared_ptr<Decryptor>()> GetColumnDataDecryptor` you are implying `CryptoContext` stores that factory rather than a `std::shared_ptr<Decryptor> data_decryptor`, so we can create a new data decryptor each time we call `InitDecryption`?

pitrou replied:
Yes, exactly!

EnricoMi replied:
I have applied that change. But something is not right. running test `file_parquet_encryption_test` fails because `GetColumnDecryptor` is now called with null `file_decryptor`.

pitrou replied:
Ah, I'll take a look.

pitrou replied:
Ok, I found the issue. When creating a factory function, we must first compute encryption parameters and store them in the closure that's returned to the caller. I took the liberty of pushing some changes to this PR, I hope you don't mind.

pitrou commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -51,8 +53,15 @@ using arrow::internal::checked_pointer_cast;
+
 // Base class to test writing and reading encrypted dataset.
-class DatasetEncryptionTestBase : public ::testing::Test {
+class DatasetEncryptionTestBase
+    : public testing::TestWithParam<std::tuple<EncryptionParam, bool>> {
```
I think this would be more readable with something like:
```c++
struct EncryptionTestParam {
  bool uniform_encryption;   // false is using per-column keys
  bool parallel;
};

class DatasetEncryptionTestBase : public testing::TestWithParam<EncryptionTestParam> {
```

pitrou reviewed (changes requested):
@EnricoMi Sorry for the delay and thanks a lot for persevering on this. I think the fix is still not 100% correct, see comments below. But we're nearing a solution.

Also, I can help on this PR if you want.

EnricoMi commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,55 @@ constexpr int32_t kBufferSizeLength = 4;
+
+ protected:
+  static inline std::function<void(EVP_CIPHER_CTX*)> ctx_deleter_ =
+      [](EVP_CIPHER_CTX* ctx) { EVP_CIPHER_CTX_free(ctx); };
```
:sparkles: smart pointer magic :sparkles:

EnricoMi replied:
Line

    using CipherContext = std::unique_ptr<EVP_CIPHER_CTX, decltype(DeleteCipherContext)>;

has to read

    using CipherContext = std::unique_ptr<EVP_CIPHER_CTX, decltype(&DeleteCipherContext)>;

With that, this works nicely.

pitrou commented:
@github-actions crossbow submit -g cpp

pitrou commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -106,19 +117,65 @@ class DatasetEncryptionTestBase : public ::testing::Test {
     auto dataset = std::make_shared<InMemoryDataset>(table_);
     EXPECT_OK_AND_ASSIGN(auto scanner_builder, dataset->NewScan());
+    // ideally, we would have UseThreads(concurrently) here, but that is not working
+    // unless GH-26818 (https://github.com/apache/arrow/issues/26818) is fixed
```
When you say "that is not working", it's simply the `AssertTablesEqual` call that would fail, right?

If that's indeed the case, we can sort the tables before comparing them, to make sure the test is not order-dependent.

EnricoMi replied:
that is correct

pitrou replied:
Ok, fixed by sorting the tables before comparison.

pitrou commented:
I think this PR is mostly looking good on the implementation side now (but would appreciate other eyes since I did some of the latest changes: @wgtmac perhaps?).

I will do a pass on the tests to polish them.

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -314,11 +314,7 @@ class SerializedFile : public ParquetFileReader::Contents {
     }
   }
 
-  void Close() override {
-    if (file_metadata_ && file_metadata_->file_decryptor()) {
-      file_metadata_->file_decryptor()->WipeOutDecryptionKeys();
-    }
-  }
```
I forget, is there a reason we removed this @EnricoMi ? The `WipeOutDecryptionKeys` method still exists.

wgtmac replied:
Same question here.

adamreeve replied:
The PR originally removed all of the "WipeOut" methods as one thread wiping out a decryptor while it was still in use was the main cause of crashes. `WipeOutDecryptionKeys` was restored after [this comment](https://github.com/apache/arrow/pull/44990/commits/c4305bc37d7ca7a44a73b2667a4a4ee02bd3e4d2#r1880648351) though. I think it's still the case that the decryption properties are shared in a multi-threaded dataset read, so we do want this code removed.

wgtmac replied:
Right, `WipeOutDecryptionKeys()` is called in the `~InternalFileDecryptor()` so we no longer need to call it from here.

> But I suspect we might see crashes again if we tested with keys set explicitly in FileDecryptionProperties.

The root cause is that a single `FileDecryptionProperties` instance may be shared by multiple Parquet reader using the same `ReaderProperties`. Should we make a copy instead of sharing?

wgtmac replied:
Right, `WipeOutDecryptionKeys()` is called in the `~InternalFileDecryptor()` so we no longer need to call it from here.

> But I suspect we might see crashes again if we tested with keys set explicitly in FileDecryptionProperties.

The root cause is that a single `FileDecryptionProperties` instance may be shared by multiple Parquet reader using the same `ReaderProperties`. Should we make a copy instead of sharing?

pitrou replied:
> Right, `WipeOutDecryptionKeys()` is called in the `~InternalFileDecryptor()` so we no longer need to call it from here.

That begs the question: when is the `InternalFileDecryptor` destroyed exactly?

adamreeve replied:
> That begs the question: when is the InternalFileDecryptor destroyed exactly?

From debugging some of the dataset tests, it looks like these are cached in the dataset, as `ParquetFileFragment` holds shared pointer to the `parquet::FileMetadata` containing the decryptor (although possibly not if `ScanOptions::cache_metadata` is disabled). So it might not be destroyed until the whole dataset is destroyed.

> Should we make a copy instead of sharing?

Maybe, although I'm not sure that would work if we re-added this wipe-out on close or weren't caching the metadata in a dataset. Eg. if you're doing a dataset scan and create a copy of the properties for each file fragment you create, the keys could be wiped out after one scan when the file is closed, and you'd need to somehow recreate a new copy or reinstate the keys if you do another scan that reuses a previously opened and closed file.

pitrou commented:
@github-actions crossbow submit -g cpp

pitrou reviewed (approved):
+1, but hopefully we can get other reviews

wgtmac commented:
Let me take a look :)

wgtmac commented on cpp/src/parquet/column_reader.h:
```diff
@@ -114,8 +115,8 @@ struct CryptoContext {
   int16_t column_ordinal = -1;
-  std::shared_ptr<Decryptor> meta_decryptor;
-  std::shared_ptr<Decryptor> data_decryptor;
+  std::function<std::shared_ptr<Decryptor>()> meta_decryptor;
```
nit: rename to `meta_decryptor_creator` or something alike for better readability.

wgtmac commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -306,8 +306,13 @@ class SerializedPageReader : public PageReader {
-  // The ordinal fields in the context below are used for AAD suffix calculation.
+  // The CryptoContext used by this PageReader.
   CryptoContext crypto_ctx_;
+  // This PageReader has its own Decryptor instances in order to be thread-safe.
```
Why not directly caching `meta_decryptor_` and `data_decryptor_` in the above `CryptoContext`?

wgtmac replied:
Yes, but isn't `CryptoContext` is created every time for a new `PageReader`? The old code has the problem because it shares decryptor instances from FileReader all the way down to PageReader. However, now we can create decryptors when creating the `CryptoContext` and cache them along with `CryptoContext`.

It is misleading that `PageReader::Open` takes `const CryptoContext*` but actually copies everything it passes.

pitrou replied:
Yeah, everything is misleading in the Parquet encryption code. It's an entire mess.

pitrou replied:
But, still, we should be careful that "copying" doesn't mean the actual objects are cloned when we've got `shared_ptr` members.

Besides, the nice thing with these factory functions is that they capture some precomputed encryption parameters such that calling them only creates a `Decryptor` instance (instead of fetching keys etc.).

wgtmac reviewed (commented):
I haven't finished reading all changes yet. Will continue tomorrow morning.

adamreeve commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,64 @@ constexpr int32_t kBufferSizeLength = 4;
+class AesCryptoContext {
+ public:
+  AesCryptoContext(ParquetCipher::type alg_id, int32_t key_len, bool metadata,
+                   bool write_length) {
```
This is used for reading and writing, so maybe `include_length` would be a better name?

adamreeve commented on cpp/src/parquet/encryption/encryption_internal.cc:
```diff
@@ -52,25 +52,64 @@ constexpr int32_t kBufferSizeLength = 4;
                             bool write_length);
 
-  ~AesEncryptorImpl() { WipeOut(); }
+  ~AesEncryptorImpl() override = default;
```
nit: This override is redundant, and similarly for `~AesDecryptorImpl`.

adamreeve reviewed (approved):
This looks good to me thanks Enrico, I've left some minor suggestions.

As I mentioned in https://github.com/apache/arrow/pull/44990#discussion_r1953634919, I think there might still be a problem with keys being wiped out if they're set explicitly rather than using a key retriever, but this is still a big improvement and that issue could be addressed separately if needed rather than blocking merging of this PR.

pitrou commented:
I pushed some more cleanups, in particular I switched to `std::unique_ptr` for non-shareable objects `Decryptor` and `AesDecryptor`.

pitrou commented:
I think at some point a large cleanup and reorganization phase would be deserved on the Parquet encryption code:
* header and source file naming
* object and method naming
* reorganizing responsabilities across objects
* wacky implementation details (e.g. why exactly do we need the mutable `CryptoContext::start_decrypt_with_dictionary_page`?)

pitrou commented:
Hmm, the Lint CI step is failing for unrelated reasons (see https://github.com/apache/arrow/issues/45521).

pitrou commented:
@github-actions crossbow submit -g cpp

raulcd commented:
@pitrou I took a moment to put your comment above into an enhancement issue here:
- https://github.com/apache/arrow/issues/45525

pitrou commented:
Oh, thank you @raulcd !

pitrou commented:
Well, there's an interesting segfault on the Windows C++ build...

EnricoMi reviewed (commented):
@pitrou quite some substantial refactoring, LGTM! Thanks for stepping in!

pitrou commented:
After several retries, it seems the Windows segfaults can happen randomly in various places, even when writing an encrypted file. Example logs (two segfaults, including one in `TestEncryptionKeyManagement.WrapLocally`):
https://github.com/pitrou/arrow/actions/runs/13314446458/job/37184921711#step:10:266

pitrou commented:
Given the symptoms, I'm not sure the Windows segfaults are directly caused by this PR's changes. I might try to reproduce on a Windows VM, though (unless someone beats me to it?).

EnricoMi commented:
This Windows segfault might be related to this PR since I cannot see it in main branch. I have checked a few recent commits and there were no segfaults retried. I can try setting up dev tools on my Windows VM this week.

EnricoMi commented:
I could not manage to set up compiling C++ Arrow on Windows in debug mode, let alone Python.

pitrou commented:
I spent a couple hours trying to build Arrow C++ on a Windows VM, but choked on Boost-related issues.

pitrou commented:
I finally managed to get a debug build running on Windows. Unfortunately, after a few attempts I cannot get the Parquet encryption tests to crash.

pitrou commented:
I'll try to continue debugging the Windows issue once https://github.com/apache/arrow/pull/45637 is merged :)

AlenkaF commented:
The PR that fixes Boost-related issues has been merged: https://github.com/apache/arrow/pull/45637 in case this might help with debugging on Wins.

Is there still a need to open a follow-up issue for "... keys being wiped out if they're set explicitly rather than using a key retriever.": https://github.com/apache/arrow/pull/44990#discussion_r1953634919?

EnricoMi commented:
@pitrou what do you suggest we do next? I see three consistent segfaults on Windows:
```
  The following tests FAILED:
  	 44 - arrow-json-integration-test (SEGFAULT)            arrow-tests unittest
  	 77 - arrow-flight-sql-test (Failed)                    arrow_flight_sql unittest
  	 81 - arrow-ipc-read-write-test (SEGFAULT)              arrow-tests unittest
```
https://github.com/EnricoMi/arrow/actions/runs/13969501352/job/39107550137#step:12:1141-1144

They don't seem to be related to Parquet encryption.

pitrou commented:
I was talking about the random segfauts that occurred in the Parquet encryption tests. If they don't occur anymore, then indeed we can probably move forward with this. But we should make sure they don't.

EnricoMi commented:
I am running C++ tests 100x here: https://github.com/EnricoMi/arrow/pull/13/files#diff-5272efdd0c7046e31936be2c954da5f68b48b5212ca4ec3ae0608f01b02aa366R102

wgtmac commented:
> After several retries, it seems the Windows segfaults can happen randomly in various places, even when writing an encrypted file. Example logs (two segfaults, including one in `TestEncryptionKeyManagement.WrapLocally`): https://github.com/pitrou/arrow/actions/runs/13314446458/job/37184921711#step:10:266

Why the CI jobs with SegFault are still green?

pitrou commented:
> > [](https://github.com/pitrou/arrow/actions/runs/13314446458/job/37184921711#step:10:266)
> 
> Why the CI jobs with SegFault are still green?

Because for some reason, at some point we decided to restart failed tests. That doesn't like a very good idea to me.

pitrou commented:
A segfault occurred again here: https://github.com/apache/arrow/actions/runs/13939766144/job/39112232657?pr=44990#step:10:267

wgtmac commented:
I will try to set up my Windows PC again to reproduce this.

EnricoMi commented:
I managed to reproduce some segfaults on Github on Linux and macOS as well by running tests 100x:
https://github.com/EnricoMi/arrow/pull/13/files#diff-5272efdd0c7046e31936be2c954da5f68b48b5212ca4ec3ae0608f01b02aa366
https://github.com/EnricoMi/arrow/actions/runs/13972966463?pr=13
So maybe we can reproduce and investigate this locally on Linux.

Here is a summary:
Fails on ARM64 AVX2 with core dump:
parquet-encryption-test [unknown Test]
https://github.com/EnricoMi/arrow/actions/runs/13972966463/job/39119355206?pr=13#step:7:23733

Fails on ARM64 with core dump:
parquet-encryption-test DecryptionTests/TestDecryptionConfiguration.TestDecryption/4
https://github.com/EnricoMi/arrow/actions/runs/13972966463/job/39119355220?pr=13#step:7:22623

Fails on macOS 13 with core dump:
parquet-encryption-key-management-test TestEncryptionKeyManagementMultiThread.WrapOnServer
https://github.com/EnricoMi/arrow/actions/runs/13972966463/job/39119355227?pr=13#step:13:19927

segfaults in other environments:
parquet-encryption-test DecryptionTests/TestDecryptionConfiguration.TestDecryption/0
parquet-encryption-key-management-test TestEncryptionKeyManagement.WrapLocally

pitrou commented:
That is great @EnricoMi . The traceback from the ARM64 core dump should help us understand what happens. I've copied it here for clarity: https://gist.github.com/pitrou/b2a936d7a2c5359e3c8b379115aab3a9

wgtmac commented:
```
diff --git a/cpp/src/parquet/file_reader.cc b/cpp/src/parquet/file_reader.cc
index 7b6c052a6..588c78558 100644
--- a/cpp/src/parquet/file_reader.cc
+++ b/cpp/src/parquet/file_reader.cc
@@ -579,11 +579,12 @@ class SerializedFile : public ParquetFileReader::Contents {
       int64_t metadata_start = read_size.first;
       metadata_len = read_size.second;
       return source_->ReadAsync(metadata_start, metadata_len)
-          .Then([this, metadata_len, is_encrypted_footer, file_decryptor](
+          .Then([this, metadata_len, is_encrypted_footer,
+                 file_decryptor = std::move(file_decryptor)](
                     const std::shared_ptr<::arrow::Buffer>& metadata_buffer) {
             // Continue and read the file footer
             return ParseMetaDataFinal(metadata_buffer, metadata_len, is_encrypted_footer,
-                                      file_decryptor);
+                                      std::move(file_decryptor));
           });
     }
     return ParseMetaDataFinal(std::move(metadata_buffer), metadata_len,
```

Should we do this?

EnricoMi commented:
> ... at some point we decided to restart failed tests. That doesn't like a very good idea to me.

@pitrou I agree, ideally, you would remove the `--repeat until-pass:3` *and* add a job that runs with `--repeat until-fail:100` to more likely discover flaky tests. Of course, after all existing flaky tests have been resolved.

EnricoMi commented:
> Should we do this?

@wgtmac Excellent catch. This is the same pattern as what I have fixed in https://github.com/apache/arrow/pull/44990/files#diff-c49cc60c1aa3a00a84b72c89ce2c8f5eb5d034ed8383741def1c4ae377d62854R529-R533. We have to `std::move` those variables that will be deconstructed when leaving the current function, when using them in a lambda (I think inside the `Then` is also required for this pattern).

I have added the outer / first `std::move`. The inner / second `std::move` is not needed as the `file_decryptor` is a `const` here / safe already, same situation as in https://github.com/apache/arrow/pull/44990#discussion_r1880772102.

EnricoMi commented on cpp/src/parquet/encryption/internal_file_decryptor.cc:
```diff
@@ -64,17 +66,20 @@ InternalFileDecryptor::InternalFileDecryptor(FileDecryptionProperties* propertie
-      aes_decryptor->WipeOut();
-    }
-  }
+  footer_key_.clear();
```
@pitrou why do we clear the footer key? Do we not trust the string destructor to wipe memory before releasing it?

The same reason probably why we wipe out keys in `properties`.

EnricoMi replied:
It looks like `std::string::clear()` sets the length of the string to 0, so the string value becomes inaccessible, but the key is not erased in the memory. No need to call `footer_key.clear()` from destructor because the string will be deleted anyways.

EnricoMi commented on cpp/src/parquet/encryption/internal_file_decryptor.cc:
```diff
@@ -64,17 +66,20 @@ InternalFileDecryptor::InternalFileDecryptor(FileDecryptionProperties* propertie
+
 void InternalFileDecryptor::WipeOutDecryptionKeys() {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::unique_lock lock(mutex_);
```
This method is only called from the destructor, so there is no need for a lock, as its guaranteed to be called only once.

wgtmac replied:
I guess the reason is that `WipeOutDecryptionKeys()` might be called in the `ParquetFileReader::Close()`

EnricoMi commented on cpp/src/parquet/encryption/internal_file_decryptor.cc:
```diff
@@ -64,17 +66,20 @@ InternalFileDecryptor::InternalFileDecryptor(FileDecryptionProperties* propertie
 void InternalFileDecryptor::WipeOutDecryptionKeys() {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::unique_lock lock(mutex_);
   properties_->WipeOutDecryptionKeys();
```
@pitrou Variable `properties` is shared across decryptor instances, wiping out keys mutates `properties` for other decryptors. Firstly, we revoke keys for other decryptors. Secondly, multiple decryptors might call `WipeOutDecryptionKeys` concurrently causing issues. A lock here does not help as individual decryptors have individual locks.

The keys should be wiped out in the destructor of `properties`, so they exist for the lifetime of `properties`.

EnricoMi commented:
Commit "Remove wipeout code, use shared_ptr instead of raw pointer" (c41246ef) consequently removes key wipe-out logic (from decrypter as well as encrypter). Properties are not mutated any more (no keys get wiped out), which makes `is_utilized`, `set_utilized` and `DeepClone` redundant. Reuse of `property` by multiple decryptors in multi-threaded environment is not an issue any more.

This seems to fix our remaining encryption segfaults: https://github.com/EnricoMi/arrow/actions/runs/14039452977

All what is left are flaky tests in `TwoLevelCacheWithExpirationTest`, no segfaults, likely race-conditions.

macOS 13
parquet-encryption-key-management-test - TwoLevelCacheWithExpirationTest.CleanupPeriodOk
https://github.com/EnricoMi/arrow/actions/runs/14039452977/job/39305637250#step:13:821

Windows 2022 AVX2
parquet-encryption-key-management-test - TwoLevelCacheWithExpirationTest.CleanupPeriodOk
https://github.com/EnricoMi/arrow/actions/runs/14039452977/job/39305637260#step:10:893

Windows MinGW64
parquet-encryption-key-management-test - TwoLevelCacheWithExpirationTest.RemoveExpiration
https://github.com/EnricoMi/arrow/actions/runs/14039452977/job/39305637284#step:12:401

EnricoMi commented:
I am seeing this `Future` related issue with `LargeRowCountEncryptionTest/LargeRowCountEncryptionTest.ReadEncryptLargeRowCount/UniformEncryptionSerial`:

    /arrow/cpp/src/arrow/util/future.cc:128:  Check failed: !IsFutureFinished(state_) Future already marked finished (SUCCESS)

This fails here (I have added the current `state_` to the error message above):
https://github.com/apache/arrow/blob/e184d7b6914ab90763c8b7e87c8baca275e14530/cpp/src/arrow/util/future.cc#L126

https://github.com/EnricoMi/arrow/actions/runs/14054489034/job/39350869592#step:7:8443
https://github.com/EnricoMi/arrow/actions/runs/14054489034/job/39350869592#step:7:8654

Any idea how this could possibly be caused by this pull request? Has any one seen this before?

wgtmac commented:
The backtrace in https://github.com/EnricoMi/arrow/actions/runs/14054489034/job/39350869592#step:7:8443 seems to point to logic in `AsyncScanner::ToTable()` and `ReadaheadGenerator::AddMarkFinishedContinuation`. 

I'm not familiar with that part. Do you have any idea? @westonpace @mapleFU

EnricoMi commented:
I have copied the full stacktrace here: https://gist.github.com/EnricoMi/16f34a971a55dcbbd7801614c888b36e

These issues / pull requests have seen the same error "Future already marked finished" before. Maybe they provide some help on how to investigate that problem:
- https://github.com/apache/arrow/issues/33519
- https://github.com/apache/arrow/issues/20442
- https://github.com/apache/arrow/issues/32430
- https://github.com/apache/arrow/issues/32408
- https://github.com/apache/arrow/issues/31691
- https://github.com/apache/arrow/issues/30094
- https://github.com/apache/arrow/pull/13880

zanmato1984 commented:
> I am seeing this `Future` related issue with `LargeRowCountEncryptionTest/LargeRowCountEncryptionTest.ReadEncryptLargeRowCount/UniformEncryptionSerial`:
> 
> ```
> /arrow/cpp/src/arrow/util/future.cc:128:  Check failed: !IsFutureFinished(state_) Future already marked finished (SUCCESS)
> ```

Hi @EnricoMi , if you can reproduce the issue fairly often, could you run it with TSAN enabled? You can refer to `ninja-debug-tsan` in `CMakePresets.json`. (I tried to reproduce it in my local but the error never occurred w/ or w/o TSAN.)

The best expectation would be TSAN warns about something and the test fails the same way - then that "something" is highly implied to be the root cause.

The second best would be TSAN is clean and the test fails the same way - then we can exclude the possibility of race.

mapleFU commented:
I'm a little suspect that `ReadaheadGenerator` don't handle the future state well, let me dive into this one

EnricoMi commented:
@mapleFU awesome finding, I can try to reproduce this in a unit test and fix this from there, if you don't plan to do this yourself. Happy to help.

mapleFU commented:
I'm testing this locally and I'll send a pr in 10minutes

mapleFU commented:
Would you mind try with this patch? https://github.com/apache/arrow/pull/45954/

zanmato1984 commented:
> I think below might be a incorrect ordering:
> 
> Thread1: check `state_->finished.load() == true` in `1` Thread2: state->MarkFinishedIfDone, state->final_future.MarkFinished Thread1: state_->num_running.fetch_add(1) in `2` Thread1: state->final_future.MarkFinished

Hmm, is this assuming two threads executing `ReadaheadGenerator::operator()()` simultaneously? I don't think this is not possible, as noted in https://github.com/apache/arrow/blob/6fee049c816fc772baee9e4f3771e592bcb87c46/cpp/src/arrow/util/async_generator.h#L45-L56

You assumption is what's-so-called "synchronous reentrant", but `ReadaheadGenerator` is only "async reentrant": there must be code somewhere (I'm not sure where) doing synchronization between `ReadaheadGenerator::operator()()` invocations from multiple threads (or the caller is just using one thread doing that). You can also see that, suppose `ReadaheadGenerator::operator()()` is actually called sync-reentrant-ly, there are more serious races in it (see the access to `readahead_queue` for example).

zanmato1984 commented:
> Would you mind try with this patch? #45954

Using lock is probably able to prevent the weird bug from happening, but I'd say it's kind of against the design purpose of async generators.

mapleFU commented:
> Hmm, is this assuming two threads executing ReadaheadGenerator::operator()() simultaneously? I don't think this is possible, as noted in

No. Thread1 is the "main execute ReadaheadGenerator::operator()() thread", and thread2 is triggered by sub-task

mapleFU commented:
> Using lock is probably able to prevent the weird bug from happening, but I'd say it's kind of against the design purpose of async generators.

An `std::atomic<int64_t>` with CAS would works in this scenerio, I just use a lock-based impl to verify it fixes this problem

pitrou commented:
> Using lock is probably able to prevent the weird bug from happening, but I'd say it's kind of against the design purpose of async generators.

I would say a lock is entirely fine, if it's only protecting a small region that does not invoke third-party code.

EnricoMi commented:
> Would you mind try with this patch? #45954

running here: https://github.com/EnricoMi/arrow/actions/runs/14102683222

zanmato1984 commented:
> No. Thread1 is the "main execute ReadaheadGenerator::operator()() thread", and thread2 is triggered by sub-task

Oh, indeed. My previous understanding was wrong. Sorry :(

> An `std::atomic<int64_t>` with CAS would works in this scenerio, I just use a lock-based impl to verify it fixes this problem.

> I would say a lock is entirely fine, if it's only protecting a small region that does not invoke third-party code.

Ah, yes this locking is alright. I was originally understanding it as locking to synchronize concurrent calls to `operator()()`. By looking again it is actually synchronizing between `operator()()` and the future continuations, which is still async-reentrant. I should've look more carefully. Sorry :(

EnricoMi commented:
Some more input:

here is the stacktrace of the thread that finishes the future successfully:
https://github.com/EnricoMi/arrow/actions/runs/14100098414/job/39494607171#step:7:8744

and this is the stacktrace of the thread that fails with `Future already marked finished (SUCCESS)`:
https://github.com/EnricoMi/arrow/actions/runs/14100098414/job/39494607171#step:7:9429

EnricoMi commented:
> > Would you mind try with this patch? #45954
> 
> running here: https://github.com/EnricoMi/arrow/actions/runs/14102683222

@mapleFU this fixes the `Future` issue, only the `TwoLevelCacheWithExpirationTest.CleanupPeriodOk` occurred across 100 test runs:
https://github.com/EnricoMi/arrow/actions/runs/14102683222/job/39502158145#step:10:972
https://github.com/EnricoMi/arrow/actions/runs/14102683222/job/39502158191#step:12:336

EnricoMi commented:
@mapleFU looks like you can confirm that the `Future` issue is not related to this pull request, so it is not blocking this?

@pitrou what are your feelings about removing the wipe-out logic completely?

EnricoMi commented:
> > @pitrou what are your feelings about removing the wipe-out logic completely?
> 
> I think each object should still wipe out its own secrets in its destructors, but it should probably not wipe out other objects.

I have reinstated clearing keys in the respective destructors. I doubt this is doing anything useful. But just in case it does, this way we don't remove this "feature".

> Would that be enough to make things work?

To make what work? I think everything is already working.

pitrou commented:
> I doubt this is doing anything useful.

Probably not, because of https://github.com/apache/arrow/issues/31603 . We would need to securely clear the keys' memory.

EnricoMi commented:
So fixing the existing wipeout logic is out-of-scope of this pull request. I'd say we keep d0dd68 to not make things worse and fix this separately.

EnricoMi commented:
Looks like all issues are resolved.

pitrou commented:
Ok, thanks a lot. I'll re-review soon.

EnricoMi commented:
@pitrou do you think it is like to get this in before the feature freeze?

AlenkaF commented:
@EnricoMi there are CMake issues that need to be fixed so I am not sure there will be capacity for the final review before the feature freeze.

pitrou commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -51,9 +53,35 @@ using arrow::internal::checked_pointer_cast;
+  bool crypto_factory;
+};
+
+std::string PrintParam(const testing::TestParamInfo<EncryptionTestParam>& info) {
```
Wouldn't it be more idiomatic to define a `operator<<` for `EncryptionTestParam`?

pitrou commented on cpp/src/arrow/dataset/file_parquet_encryption_test.cc:
```diff
@@ -51,9 +53,35 @@ using arrow::internal::checked_pointer_cast;
+struct EncryptionTestParam {
+  bool uniform_encryption;  // false is using per-column keys
+  bool concurrently;
+  bool crypto_factory;
```
Nit: call this `use_crypto_factory`?

pitrou commented on cpp/src/parquet/file_reader.cc:
```diff
@@ -638,11 +635,12 @@ class SerializedFile : public ParquetFileReader::Contents {
 
-  std::string HandleAadPrefix(FileDecryptionProperties* file_decryption_properties,
-                              EncryptionAlgorithm& algo);
+  std::string HandleAadPrefix(
+      const std::shared_ptr<FileDecryptionProperties>& file_decryption_properties,
+      EncryptionAlgorithm& algo);
```
Ok, can we take the opportunity to change the mutable ref argument to const-ref?
```suggestion
  std::string HandleAadPrefix(
      const std::shared_ptr<FileDecryptionProperties>& file_decryption_properties,
      const EncryptionAlgorithm& algo);
```

pitrou reviewed (commented):
This looks good overall to me, here are some minor comments.

Also can you rebase/merge from main to get a greener CI?

EnricoMi commented:
@pitrou suggestions applied and rebased

pitrou commented:
@github-actions crossbow submit -g cpp -g python

pitrou commented:
@github-actions crossbow submit *cp313*

pitrou commented:
@github-actions crossbow submit -g cpp

pitrou reviewed (approved):
LGTM. Thanks a lot for this @EnricoMi !

EnricoMi commented:
@pitrou thank you for your time! Will this make it into the 20.0.0 release?

pitrou commented:
@EnricoMi It should, I'll wait for CI and merge if green.

zanmato1984 commented:
I think the release 20.0.0 has been feature-freezed, and the maint-20.0.0 has been branched earlier today?

EnricoMi commented:
Can we consider this a bug fix and not a feature so it could still make it into the 20.0.0 maint branch? Or is "feature freeze" more a general "code freeze"?

pitrou commented:
This is definitely a bug fix (it fixes an actual bug) even though it's quite elaborate and goes deeper than simply preventing nasty behavior.

zanmato1984 commented:
Well, I mean that with feature freeze done (== maintenance branch branched out), it doesn't happen naturally for this PR to be in 20.0.0. We may ask for help from release managers. And yes a bug fix still gets a chance to be back-ported (depending on the severity I guess?).

raulcd commented:
The release manager (@assignUser ) has been notified and is happy to add it to the maintenance branch, as per Zulip convesation.

pitrou commented:
CI failures are unrelated.

EnricoMi commented:
Awesome, many thanks! Happy to help with the back-porting, please let me know if I can help!
